cbmc/proofs/mqtt_state.c

```diff
@@ -0,0 +1,698 @@
+#include "iot_config.h"
+#include "private/iot_mqtt_internal.h"
+
+#include <stdlib.h>
+
+#include "mqtt_state.h"
+
+/****************************************************************
+ * Bound the lengths of subscription and operation lists.
+ *
+ * Lists are empty by defaut.
+ * Set this variable to L+1 for lists of length <= L.
+ ****************************************************************/
+
+#ifndef SUBSCRIPTION_COUNT_MAX
+  #define SUBSCRIPTION_COUNT_MAX 1
+#endif
+
+#ifndef OPERATION_COUNT_MAX
+  #define OPERATION_COUNT_MAX 1
+#endif
+
+/****************************************************************
+ * Model a malloc that can fail and return NULL.
+ ****************************************************************/
+
+void *malloc_can_fail( size_t size )
+{
+  __CPROVER_assert( VALID_CBMC_SIZE( size ), "malloc_can_fail size too big" );
+  return nondet_bool() ? NULL : malloc( size );
+}
+
+/****************************************************************
+ * Model an opaque type as nothing but a valid pointer.
+ ****************************************************************/
+
+void *allocate_opaque_type()
+{
+  return malloc( 1 ); // consider using malloc(0)
+}
+
+/****************************************************************
+ * MqttOperation
+ ****************************************************************/
+
+IotMqttOperation_t allocate_IotMqttOperation( IotMqttOperation_t pOp,
+					      IotMqttConnection_t pConn )
+{
+  // assume a valid connection for the operation
+  assert( pConn != NULL );
+
+  if ( pOp == NULL ) pOp = malloc_can_fail( sizeof( *pOp ) );
+  if ( pOp == NULL ) return NULL;
+
+  pOp->pMqttConnection = pConn;
+
+  if ( pOp->incomingPublish ) {
+    // allocate publish member of the union
+
+    allocate_IotMqttPublishInfo( &(pOp->u.publish.publishInfo ) );
+
+    size_t length;
+    __CPROVER_assume( VALID_CBMC_SIZE( length ) );
+    pOp->u.publish.pReceivedData = malloc_can_fail( length );
```

Comment:
Not sure if relevant, but the members pOp->u.publish.publishInfo.pTopicName and pOp->u.publish.publishInfo.pPayload will always point to locations in this pReceivedData buffer.  Their combined lengths will also be smaller than the length of pReceivedData (i.e. both will be contained within pReceivedData).